src/main/scala/com/campudus/tableaux/controller/MediaController.scala

```diff
@@ -0,0 +1,135 @@
+package com.campudus.tableaux.controller
+
+import java.util
+import java.util.UUID
+
+import com.campudus.tableaux.TableauxConfig
+import com.campudus.tableaux.database.domain._
+import com.campudus.tableaux.database.model.FolderModel.FolderId
+import com.campudus.tableaux.database.model.{FileModel, FolderModel}
+import com.campudus.tableaux.helper.FutureUtils
+import com.campudus.tableaux.router.UploadAction
+
+import scala.concurrent.{Promise, Future}
+import scala.reflect.io.Path
+import scala.util.Failure
+
+object MediaController {
+  def apply(config: TableauxConfig, folderModel: FolderModel, fileModel: FileModel): MediaController = {
+    new MediaController(config, folderModel, fileModel)
+  }
+}
+
+class MediaController(override val config: TableauxConfig,
+                      override protected val repository: FolderModel,
+                      protected val fileModel: FileModel) extends Controller[FolderModel] {
+
+  import FutureUtils._
+
+  lazy val uploadsDirectory = Path(s"${config.workingDirectory}/${config.uploadsDirectory}")
+
+  def retrieveFolder(id: FolderId): Future[ExtendedFolder] = {
+    for {
+      folder <- repository.retrieve(id)
+      subfolders <- repository.retrieveSubfolders(id)
+      files <- fileModel.retrieveFromFolder(id)
+    } yield ExtendedFolder(folder, subfolders, files)
+  }
+
+  def addNewFolder(name: String, description: String, parent: Option[FolderId]): Future[Folder] = {
+    repository.add(Folder(name, description, parent))
+  }
+
+  def changeFolder(id: FolderId, name: String, description: String, parent: Option[FolderId]): Future[Folder] = {
+    repository.update(Folder(Some(id), name, description, parent, None, None))
+  }
+
+  def deleteFolder(id: FolderId): Future[Folder] = {
+    for {
+      folder <- repository.retrieve(id)
+
+      // delete files
+      _ <- deleteFilesOfFolder(id)
+
+      // delete subfolders
+      _ <- deleteSubfolders(id)
```

Comment:
Can we use `Future.zip` on `deleteFilesOfFolder(id)` and `deleteSubfolders(id)`? That would make both futures happen at the same time and not waiting for the files to be done and then wait for the subfolders to be deleted.